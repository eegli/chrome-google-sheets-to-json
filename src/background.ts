chrome.runtime.onInstalled.addListener(() => {
  // Page actions are disabled by default and enabled on select tabs
  //@ts-ignore the typings seem to be very alpha stagey
  chrome.action.disable();

  // Clear all rules to ensure only our expected rules are set
  chrome.declarativeContent.onPageChanged.removeRules(undefined, () => {
    // Declare a rule to enable the action on example.com pages
    let exampleRule = {
      conditions: [
        new chrome.declarativeContent.PageStateMatcher({
          pageUrl: { hostEquals: 'docs.google.com' },
        }),
        new chrome.declarativeContent.PageStateMatcher({
          pageUrl: { pathContains: 'spreadsheets/d' },
        }),
      ],
      actions: [new chrome.declarativeContent.ShowPageAction()],
    };

    // Finally, apply our new array of rules
    let rules = [exampleRule];
    chrome.declarativeContent.onPageChanged.addRules(rules);
  });
});
