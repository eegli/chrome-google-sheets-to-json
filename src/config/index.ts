export default {
  defaultSheet: '1',
};
