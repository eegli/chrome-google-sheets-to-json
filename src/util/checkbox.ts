export function insertCheckboxes(parent: HTMLElement, data: string[]): void {
  let idx = data.length - 1;
  // Looping backwards through the sheet names and appending a new node before the first child of the parent

  for (; idx >= 0; idx--) {
    const el = data[idx];
    const container = document.createElement('div');
    const input = document.createElement('input');
    const label = document.createElement('label');

    input.type = 'radio';
    input.name = 'sheetSelect';
    input.value = el;
    input.id = el;
    if (idx === 0) input.checked = true;

    label.htmlFor = el;
    label.innerHTML = el;

    container.classList.add('form-check');
    input.classList.add('form-check-input');
    label.classList.add('form-check-label');

    container.append(input);
    container.append(label);

    parent.prepend(container);
  }
}
