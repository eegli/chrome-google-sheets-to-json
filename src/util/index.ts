import * as GST from 'google-spreadsheets-ts';

type DownloadArgs = {
  url: string;
  page: number;
  fileName: string;
};

export async function downloadJSON(opts: DownloadArgs): Promise<void> {
  const { url, page, fileName } = opts;
  const JSONendpoint = getJSONEndpoint(url, page);
  try {
    // Fetch data, extract and prompt download
    const response = await fetch(JSONendpoint);
    const data: GST.RootObject = await response.json();
    const json = extractJSON(data);
    const jsonStr = JSON.stringify(json);
    const blob = new Blob([jsonStr], { type: 'application/json' });
    const downloadUrl = URL.createObjectURL(blob);

    chrome.downloads.download({
      url: downloadUrl,
      filename: fileName + '.json'
    });
  } catch (e) {
    throw new Error('Error downloading JSON');
  }
}

function getJSONEndpoint(docsUrl: string, page: number): string {
  const [sheetId] = docsUrl.match('(?<=\\/d\\/)[^\\/]*') || [''];
  return `https://spreadsheets.google.com/feeds/list/${sheetId}/${page}/public/values?alt=json`;
}

function extractJSON(data: GST.RootObject): any[] {
  const res: Record<string, any>[] = [];

  const rows = data.feed.entry;

  // Empty Google Sheet
  if (!rows) return res;

  for (const row of rows) {
    const formattedRow = {} as Record<string, any>;

    for (const key in row) {
      if (key.startsWith('gsx$') && !key.endsWith('_cpzh4')) {
        formattedRow[key.replace('gsx$', '')] = row[key].$t;
      }
    }

    res.push(formattedRow);
  }

  return res;
}
