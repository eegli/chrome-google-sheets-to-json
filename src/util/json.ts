import * as GST from 'google-spreadsheets-ts';

export function getJSONEndpoint(id: string, page: number): string {
  return `https://spreadsheets.google.com/feeds/list/${id}/${page}/public/values?alt=json`;
}
export function extractJSON(data: GST.RootObject): any[] {
  const res: Record<string, any>[] = [];

  const rows = data.feed.entry;

  // Empty Google Sheet
  if (!rows) return res;

  for (const row of rows) {
    const formattedRow = {} as Record<string, any>;

    for (const key in row) {
      if (key.startsWith('gsx$') && !key.endsWith('_cpzh4')) {
        formattedRow[key.replace('gsx$', '')] = row[key].$t;
      }
    }

    res.push(formattedRow);
  }

  return res;
}
