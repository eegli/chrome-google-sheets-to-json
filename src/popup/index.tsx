import { FunctionalComponent, render } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import { sheetNamesContentScript } from '../scripts/sheet';
import { downloadJSON } from '../util';
import CheckBoxes from './checkbox';

export type Sheets = string[];

const Popup: FunctionalComponent = () => {
  const [activeTab, setActiveTab] = useState<chrome.tabs.Tab>();
  const [sheets, setSheets] = useState<Sheets>([]);
  const [selectedSheet, setSelectedSheet] = useState<string>('');
  const [isPublicDoc, setIsPublicDoc] = useState<boolean>(true);

  /* https://preactjs.com/guide/v10/typescript/#typing-events */
  function handleSelectSheet(event: Event) {
    event.preventDefault();
    if (event.target instanceof HTMLInputElement) {
      setSelectedSheet(event.target.value);
    }
  }

  function handleSubmit(event: Event): void {
    event.preventDefault();
    if (event.target instanceof HTMLFormElement) {
      handleDownload();
    }
  }

  async function handleDownload(): Promise<void> {
    // Each sheet has its own url. The first sheet has number 1, the
    // second number 2, etc
    const page = sheets.indexOf(selectedSheet) + 1;
    const url = activeTab?.url || '';
    const fileName =
      activeTab?.title?.replace(' - Google Sheets', '') || 'data';

    // Download JSON
    try {
      await downloadJSON({ url, page, fileName });
      window.close();
    } catch (e) {
      console.log('Sheet is not public');
      console.error('An error occured', e);
      setIsPublicDoc(false);
    }
  }

  // https://dev.to/stlnick/useeffect-and-async-4da8
  useEffect(() => {
    (async () => {
      const [tab] = await chrome.tabs.query({
        active: true,
        currentWindow: true
      });

      setActiveTab(tab);

      const sheets = await chrome.scripting.executeScript({
        target: { tabId: tab.id || 0 },
        func: sheetNamesContentScript
      });

      const sheetNames: string[] = sheets[0].result;

      setSheets(sheetNames);
      setSelectedSheet(sheetNames[0]);
    })();
  }, []);

  return (
    <main>
      <div className="container w-auto p-3">
        {isPublicDoc ? (
          <div>
            <h4 className="text-nowrap">Choose Sheet</h4>
            <form name="sheet" onSubmit={handleSubmit}>
              <CheckBoxes
                sheets={sheets}
                selectedSheet={selectedSheet}
                onChange={handleSelectSheet}
              />
              <button type="submit" className="btn btn-primary my-3">
                Submit
              </button>
            </form>
          </div>
        ) : (
          <div className="alert alert-danger" role="alert">
            <h4 className="alert-heading text-nowrap">Sheet is not public</h4>
            <p>
              It looks like the Sheet "{selectedSheet}" is not published. As of
              now, this extension can only download Google Sheets that are
              published to the web.
            </p>
            <hr />
            <p className="mb-0">
              Publish your Sheet and try again.
              <br />
              <a
                href="https://support.google.com/docs/answer/183965"
                rel="noopener noreferrer"
                target="_blank"
                class="alert-link">
                How to publish a Google Sheet
              </a>
            </p>
          </div>
        )}
      </div>
    </main>
  );
};

render(<Popup />, document.body);
