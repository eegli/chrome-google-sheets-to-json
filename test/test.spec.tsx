function add(num: number) {
  return num + 1;
}

import sampleJSON from './sample-json.json';
import Popup from '../src/popup/error';
import { extractJSON } from '../src/util';

describe('Test', () => {
  it('works', () => {
    expect(add(1)).toEqual(2);
  });
  it('matches snapshot', () => {
    expect(<Popup sheetName="testSheet" />).toMatchSnapshot();
  });
  it('extracts JSON', () => {
    const json = extractJSON(sampleJSON);
    expect(json).toMatchSnapshot();
  });
});
