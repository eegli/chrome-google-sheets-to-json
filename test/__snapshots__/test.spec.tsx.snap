// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test extracts JSON 1`] = `
Array [
  Object {
    "arena": "",
    "capacity": "",
    "founded": "",
    "joinedleague": "",
    "location": "City",
    "team": "",
  },
  Object {
    "arena": "Raiffeisen Arena",
    "capacity": "4,600",
    "founded": "1973",
    "joinedleague": "2021",
    "location": "Porrentruy",
    "team": "*HC Ajoie*",
  },
  Object {
    "arena": "Nuova Valascia",
    "capacity": "6,775",
    "founded": "1937",
    "joinedleague": "1985",
    "location": "Ambrì",
    "team": "*HC Ambrì-Piotta*",
  },
  Object {
    "arena": "PostFinance Arena",
    "capacity": "17,031",
    "founded": "1931",
    "joinedleague": "1986",
    "location": "Bern",
    "team": "*SC Bern*",
  },
  Object {
    "arena": "Tissot Arena",
    "capacity": "6,521",
    "founded": "1939",
    "joinedleague": "2008",
    "location": "Biel/Bienne",
    "team": "*EHC Biel*",
  },
  Object {
    "arena": "Vaillant Arena",
    "capacity": "6,800",
    "founded": "1921",
    "joinedleague": "1993",
    "location": "Davos",
    "team": "*HC Davos*",
  },
  Object {
    "arena": "BCF Arena",
    "capacity": "8,934",
    "founded": "1938",
    "joinedleague": "1980",
    "location": "Fribourg",
    "team": "*Fribourg-Gottéron*",
  },
  Object {
    "arena": "Patinoire des Vernets",
    "capacity": "7,135",
    "founded": "1905",
    "joinedleague": "2001",
    "location": "Geneva",
    "team": "*Genève-Servette HC*",
  },
  Object {
    "arena": "Vaudoise Aréna",
    "capacity": "9,600",
    "founded": "1922",
    "joinedleague": "2013",
    "location": "Lausanne",
    "team": "*Lausanne HC*",
  },
  Object {
    "arena": "Cornèr Arena",
    "capacity": "7,800",
    "founded": "1941",
    "joinedleague": "1981",
    "location": "Lugano",
    "team": "*HC Lugano*",
  },
  Object {
    "arena": "St. Galler Kantonalbank Arena",
    "capacity": "6,200",
    "founded": "1945",
    "joinedleague": "2018",
    "location": "Rapperswil-Jona",
    "team": "*SC Rapperswil-Jona Lakers*",
  },
  Object {
    "arena": "Ilfis Stadium",
    "capacity": "6,000",
    "founded": "1946",
    "joinedleague": "2015",
    "location": "Langnau im Emmental",
    "team": "*SCL Tigers*",
  },
  Object {
    "arena": "Hallenstadion",
    "capacity": "11,200",
    "founded": "1930",
    "joinedleague": "1989",
    "location": "Zürich",
    "team": "*ZSC Lions*",
  },
  Object {
    "arena": "Bossard Arena",
    "capacity": "7,200",
    "founded": "1967",
    "joinedleague": "1987",
    "location": "Zug",
    "team": "*EV Zug*",
  },
]
`;

exports[`Test matches snapshot 1`] = `
Object {
  "__": null,
  "__b": 0,
  "__c": null,
  "__d": undefined,
  "__e": null,
  "__h": null,
  "__k": null,
  "__self": undefined,
  "__source": undefined,
  "__v": -1,
  "constructor": undefined,
  "key": undefined,
  "props": Object {
    "sheetName": "testSheet",
  },
  "ref": undefined,
  "type": [Function],
}
`;
